### ISSUE TITLE
`probs.sample()` affects reproducibility of the gradient.

### ISSUE THREAD

**Issue by vwxyzjn** [pytorch/pytorch#23817]
## 🐛 Bug

The mere execution of `torch.distributions.categorical.Categorical.sample()` without even using its resulting tensor in the loss function seems to have an effect on the gradient calculation. 

## To Reproduce

1. Download the file at https://gist.github.com/vwxyzjn/32e0f618d97432541863236bf012fed5/raw/ec92d896e55fc9f97b401121c174e147528400ec/debugppo.py
1. Run `pip install gym tensorboardX`
1. Run `python debugppo.py`

In the following file, I have been working on a copy of PPO (proximal policy gradient) algorithm. I have seeded the environment to the best I could 
```python
# Reference: http://inoryy.com/post/tensorflow2-deep-reinforcement-learning/

import torch
import torch.nn as nn
import torch.optim as optim
import torch.nn.functional as F
from torch.distributions.categorical import Categorical
from tensorboardX import SummaryWriter

import argparse
import numpy as np
import gym
import time
import random

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='PPO agent')
    # Common arguments
    parser.add_argument('--exp-name', type=str, default="ppo",
                       help='the name of this experiment')
    parser.add_argument('--gym-id', type=str, default="CartPole-v0",
                       help='the id of the gym environment')
    parser.add_argument('--learning-rate', type=float, default=7e-4,
                       help='the learning rate of the optimizer')
    parser.add_argument('--seed', type=int, default=5,
                       help='seed of the experiment')
    parser.add_argument('--episode-length', type=int, default=200,
                       help='the maximum length of each episode')
    parser.add_argument('--total-timesteps', type=int, default=50000,
                       help='total timesteps of the experiments')
    parser.add_argument('--torch-deterministic', type=bool, default=True,
                       help='whether to set `torch.backends.cudnn.deterministic=True`')
    parser.add_argument('--prod-mode', type=bool, default=False,
                       help='run the script in production mode and use wandb to log outputs')
    
    # Algorithm specific arguments
    parser.add_argument('--gamma', type=float, default=0.99,
                       help='the discount factor gamma')
    parser.add_argument('--vf-coef', type=float, default=0.25,
                       help="value function's coefficient the loss function")
    parser.add_argument('--max-grad-norm', type=float, default=0.5,
                       help='the maximum norm for the gradient clipping')
    parser.add_argument('--ent-coef', type=float, default=0.01,
                       help="policy entropy's coefficient the loss function")
    parser.add_argument('--clip-coef', type=float, default=0.2,
                       help="the surrogate clipping coefficient")
    parser.add_argument('--update-frequency', type=int, default=3,
                        help="the frequency to update the policy network")
    args = parser.parse_args()
    if not args.seed:
        args.seed = int(time.time())

# TRY NOT TO MODIFY: setup the environment
env = gym.make(args.gym_id)
random.seed(args.seed)
np.random.seed(args.seed)
torch.manual_seed(args.seed)
torch.backends.cudnn.deterministic = args.torch_deterministic
env.seed(args.seed)

# TODO: initialize agent here:
class Policy(nn.Module):
    def __init__(self):
        super(Policy, self).__init__()
        self.fc1 = nn.Linear(4, 120)
        self.fc2 = nn.Linear(120, 84)
        self.fc3 = nn.Linear(84, 2)

    def forward(self, x):
        x = torch.Tensor(x).float().view(torch.Tensor(x).shape[0], -1)
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = self.fc3(x)
        return x

class Value(nn.Module):
    def __init__(self):
        super(Value, self).__init__()
        self.fc1 = nn.Linear(4, 64)
        self.fc2 = nn.Linear(64, 1)

    def forward(self, x):
        x = torch.Tensor(x).float().view(torch.Tensor(x).shape[0], -1)
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        return x

pg = Policy()
vf = Value()
optimizer = optim.Adam(list(pg.parameters()) + list(vf.parameters()), lr=args.learning_rate)
loss_fn = nn.MSELoss()

# TRY NOT TO MODIFY: start the game
experiment_name = f"{time.strftime('%b%d_%H-%M-%S')}__{args.exp_name}__{args.seed}"
writer = SummaryWriter(f"runs/{experiment_name}")
writer.add_text('hyperparameters', "|param|value|\n|-|-|\n%s" % (
        '\n'.join([f"|{key}|{value}|" for key, value in vars(args).items()])))
if args.prod_mode:
    import wandb
    wandb.init(project="cleanRL", tensorboard=True, config=vars(args), name=experiment_name)
    writer = SummaryWriter(f"/tmp/{experiment_name}")
global_step = 0
while global_step < args.total_timesteps:
    next_obs = np.array(env.reset())
    actions = np.empty((args.episode_length,), dtype=object)
    rewards, dones = np.zeros((2, args.episode_length))
    obs = np.empty((args.episode_length,) + env.observation_space.shape)
    
    # TODO: put other storage logic here
    values = torch.zeros((args.episode_length))
    neglogprobs = torch.zeros((args.episode_length,))
    entropys = torch.zeros((args.episode_length,))
    
    # TRY NOT TO MODIFY: prepare the execution of the game.
    for step in range(args.episode_length):
        global_step += 1
        obs[step] = next_obs.copy()
        
        # TODO: put action logic here
        logits = pg.forward([obs[step]])
        values[step] = vf.forward([obs[step]])
        probs = Categorical(logits=logits)
        action = probs.sample()
        actions[step] = action.tolist()[0]
        neglogprobs[step] = -probs.log_prob(action)
        entropys[step] = probs.entropy()
        
        # TRY NOT TO MODIFY: execute the game and log data.
        next_obs, rewards[step], dones[step], _ = env.step(actions[step])
        next_obs = np.array(next_obs)
        if dones[step]:
            break
    
    # TODO: training.
    # calculate the discounted rewards, or namely, returns
    returns = np.zeros_like(rewards)
    for t in reversed(range(rewards.shape[0]-1)):
        returns[t] = rewards[t] + args.gamma * returns[t+1] * (1-dones[t])
    # advantages are returns - baseline, value estimates in our case
    advantages = returns - values.detach().numpy()

    neglogprobs = neglogprobs.detach()
    non_empty_idx = np.argmax(dones) + 1
    for _ in range(args.update_frequency):
        logits = pg.forward(obs[:non_empty_idx])
        
        ###########################
        #
        probs.sample() 
        #
        # This seems to be a bug with pytorch
        # if you comment the line above out, you get different gradients 
        # for some reason...
        ###########################
        
        action = torch.LongTensor(actions[:non_empty_idx].astype(np.int))
        new_neglogprobs =  -probs.log_prob(action)
        ratio = torch.exp(neglogprobs[:non_empty_idx] - new_neglogprobs)
        surrogate1 = ratio * torch.Tensor(advantages)[:non_empty_idx]
        surrogate2 = torch.clamp(ratio, 1-args.clip_coef, 1+args.clip_coef) * torch.Tensor(advantages)[:non_empty_idx]
        clip = torch.min(surrogate1, surrogate2)
        vf_loss = loss_fn(torch.Tensor(returns), torch.Tensor(values)) * args.vf_coef
        loss = vf_loss - (clip + entropys[:non_empty_idx] * args.ent_coef).mean()
        
        optimizer.zero_grad()
        loss.backward(retain_graph=True)
        nn.utils.clip_grad_norm_(list(pg.parameters()) + list(vf.parameters()), args.max_grad_norm)
        optimizer.step()
        print(actions[:non_empty_idx])
        print(pg.fc1.weight.grad[:4])

    # TRY NOT TO MODIFY: record rewards for plotting purposes
    writer.add_scalar("charts/episode_reward", rewards.sum(), global_step)
env.close()
```
but if I comment out `probs.sample()` on line 151, I get completely different results. 

If you don't comment out the `probs.sample() `, then the printed gradients and actions looks like the following 
```
[1 0 0 0 1 0 1 0 1 1 1 0 1 0 0 0 0 1 1 1 0 0 1]
tensor([[ 0.0007,  0.0027, -0.0012, -0.0071],
        [-0.0007, -0.0025,  0.0011,  0.0066],
        [-0.0000,  0.0000,  0.0000,  0.0000],
        [-0.0000,  0.0000,  0.0000,  0.0000]])
[1 0 0 0 1 0 1 0 1 1 1 0 1 0 0 0 0 1 1 1 0 0 1]
tensor([[ 0.0008,  0.0029, -0.0013, -0.0076],
        [-0.0006, -0.0023,  0.0010,  0.0061],
        [ 0.0000,  0.0000,  0.0000,  0.0000],
        [ 0.0000,  0.0000,  0.0000,  0.0000]])
[1 0 0 0 1 0 1 0 1 1 1 0 1 0 0 0 0 1 1 1 0 0 1]
tensor([[ 0.0008,  0.0031, -0.0014, -0.0080],
        [-0.0006, -0.0022,  0.0010,  0.0056],
        [ 0.0000,  0.0000,  0.0000,  0.0000],
        [ 0.0000,  0.0000,  0.0000,  0.0000]])
[0 0 0 1 0 1 1 1 1 1 1 1 0 1 0 0 1 0 0 0 1 0 1 0 1 1 0 0 1 0 1 0 1 0 0 1 1
 1 0]
tensor([[ 5.3545e-05,  2.9391e-04,  1.6467e-04,  3.2634e-04],
        [ 1.0998e-08, -9.2959e-08,  5.9564e-08,  4.1972e-07],
        [-3.2921e-10,  4.9592e-09,  6.3401e-10, -5.4726e-09],
        [ 0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00]])
[0 0 0 1 0 1 1 1 1 1 1 1 0 1 0 0 1 0 0 0 1 0 1 0 1 1 0 0 1 0 1 0 1 0 0 1 1
 1 0]
tensor([[ 5.6244e-05,  3.0872e-04,  1.7297e-04,  3.4279e-04],
        [ 9.7676e-09, -8.6699e-08,  5.4406e-08,  3.8535e-07],
        [-3.2805e-10,  4.9418e-09,  6.3177e-10, -5.4533e-09],
        [ 0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00]])
```
while if you do comment out, you get a different gradient
```
[1 0 0 0 1 0 1 0 1 1 1 0 1 0 0 0 0 1 1 1 0 0 1]
tensor([[ 0.0007,  0.0027, -0.0012, -0.0071],
        [-0.0007, -0.0025,  0.0011,  0.0066],
        [-0.0000,  0.0000,  0.0000,  0.0000],
        [-0.0000,  0.0000,  0.0000,  0.0000]])
[1 0 0 0 1 0 1 0 1 1 1 0 1 0 0 0 0 1 1 1 0 0 1]
tensor([[ 0.0008,  0.0029, -0.0013, -0.0076],
        [-0.0006, -0.0023,  0.0010,  0.0061],
        [ 0.0000,  0.0000,  0.0000,  0.0000],
        [ 0.0000,  0.0000,  0.0000,  0.0000]])
[1 0 0 0 1 0 1 0 1 1 1 0 1 0 0 0 0 1 1 1 0 0 1]
tensor([[ 0.0008,  0.0031, -0.0014, -0.0080],
        [-0.0006, -0.0022,  0.0010,  0.0056],
        [ 0.0000,  0.0000,  0.0000,  0.0000],
        [ 0.0000,  0.0000,  0.0000,  0.0000]])
[1 1 1 0 0 0 1 0 1 1 1 1 1 1 1 0 1]
tensor([[-1.6087e-10, -1.4047e-09,  3.4414e-10,  4.5244e-10],
        [ 0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00],
        [-1.4134e-03, -1.3017e-02,  2.1393e-03,  2.0959e-02],
        [ 0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00]])
[1 1 1 0 0 0 1 0 1 1 1 1 1 1 1 0 1]
tensor([[-1.6798e-10, -1.4667e-09,  3.5934e-10,  4.7242e-10],
        [ 0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00],
        [-1.4333e-03, -1.3200e-02,  2.1693e-03,  2.1254e-02],
        [ 0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00]])
```

As shown in the file above, the resulting tensor of `probs.sample()` is not even stored or used in the calculation of the loss function. How its mere execution could affect the gradient calculation is a little beyond me. 

## Expected behavior

The gradient calculation and the resulting actions array really should have been the same.

## Environment

PyTorch version: 1.1.0
Is debug build: No
CUDA used to build PyTorch: 10.0

OS: Microsoft Windows 10 Pro
GCC version: Could not collect
CMake version: Could not collect

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: GPU 0: GeForce GTX 970M
Nvidia driver version: 411.31
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.16.2
[pip] numpydoc==0.8.0
[pip] torch==1.1.0
[pip] torchvision==0.2.2.post3
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.3                      203  
[conda] mkl-service               1.1.2            py37hb782905_5  
[conda] mkl_fft                   1.0.10           py37h14836fe_0  
[conda] mkl_random                1.0.2            py37h343c172_0  
[conda] torch                     1.1.0                    pypi_0    pypi
[conda] torchvision               0.2.2.post3              pypi_0    pypi

**Comment by SsnL** [comment]
It appears to me that you don't get the same `non_empty_idx`. Since `.sample` advances rng, it is very possible that the sampled action leads to a different outcome and thus different loss and different gradients.

On an orthogonal note, it isn't immediately clear to me why you need `retain_graph=True`. If you do not intend to reuse graph, some logic may be implemented incorrectly

**Comment by vwxyzjn** [comment]
Hi @SsnL,

Thank you so much for the prompt response. I think the rng advances explanation makes sense. Just now I was trying to run the algorithm again with multiple different seeds, and it appears the `seed=5` I was experimenting with was the only one that has a flat and degenerate performance for about 18k timesteps as shown below. Maybe it was just unlucky.

![image](https://user-images.githubusercontent.com/5555347/62494621-efa9a180-b7a1-11e9-9d53-2ae05900f909.png)

Regarding your kind note, the PPO maximizes the following objective:

![image](https://user-images.githubusercontent.com/5555347/62494884-73638e00-b7a2-11e9-9848-c0e9667cd44f.png)

where 

![image](https://user-images.githubusercontent.com/5555347/62494969-ac9bfe00-b7a2-11e9-8a94-7ff08665f64b.png)

If you think about it, the `clip(r_t(theta), 1+e, 1-e)`, which is used in the loss calculation, is always going to be 1 when you run the backpropagation for the first time because `pi_theta` and `pi_theta_old` is the same. It is only that you *retain* the graph and run the optimization again that the `clip(r_t(theta), 1+e, 1-e)` makes sense. That's why the authors suggest " optimize it with minibatch SGD for K epochs“, which, as I understood it, requires retainment of graph for doing the optimization for the second time.
